File: python/eos/van_der_waals_eos.py
from eos import CubicEosBase
from scipy.constants import gas_constant
import numpy as np


class IsobaricIsothermalState:
    """ Constant pressure-temperature state for PR EoS """

    def __init__(self, A, B):
        """
        Parameters
        ----------
        A : float
            Reduced attraction parameter
        B : float
            Reduced repulsion parameter
        """
        self.__A = A
        self.__B = B

    @property
    def reduced_attraction_param(self):
        return self.__A

    @property
    def reduced_repulsion_param(self):
        return self.__B

class VanDerWaalsEos(CubicEosBase):
    """
    Van der Waals equation of state.
    """
    __OMEGA_A = 0.421875
    __OMEGA_B = 0.125

    def __init__(self, pc, tc):
        """
        Parameters
        ----------
        pc : float
            Critical pressure
        tc : float
            Critical temperature
        """
        super().__init__(pc, tc, self.__OMEGA_A, self.__OMEGA_B)

    @staticmethod
    def _zfactor_cubic_eq(A, B):
        """
        Computes coefficients of the cubic equation of Z-factor.

        Paramters
        ---------
        A : float
            Reduced attraction parameter
        B : float
            Reduced repulsion parameter

        Returns
        -------
        list
            Coefficients of the cubic equation
        """
        return [1.0, -B - 1.0, A, -A*B]

    @staticmethod
    def _ln_fugacity_coeff_impl(z, A, B):
        """
        Computes the natural log of fugacity coefficient

        Parameters
        ----------
        z : float
            Z-factor
        A : float
            Reduced attraction parameter
        B : float
            Reduced repulsion parameter

        Returns
        -------
        float
            Natural log of fugacity coefficient
        """
        return -np.log(z - B) - A/z + z - 1.0

    def pressure(self, t, v):
        """
        Computes pressure at a given volume and temperature.

        Parameters
        ----------
        t : float
            Temperature
        v : array_like
            Volume

        Returns
        -------
        array_like
            Pressure
        """
        a = self.attraction_param
        b = self.repulsion_param
        return gas_constant*t/(v - b) - a/v**2

    def create_state(self, p, t):
        """
        Creates constant pressure-temperature state

        Parameters
        ----------
        p : float
            Pressure
        t : float
            Temperature

        Returns
        -------
        IsobaricIsothermalState
        """
        A = self._calc_reduced_attraction_param(p, t)
        B = self._calc_reduced_repulsion_param(p, t)
        return IsobaricIsothermalState(A, B)

    @staticmethod
    def zfactors(state):
        """
        Computes Z-factor

        Parameters
        ----------
        state : IsobaricIsothermalState

        Returns
        -------
        numpy.array
            Z-factors in the ascending order
        """
        A = state.reduced_attraction_param
        B = state.reduced_repulsion_param
        return CubicEosBase.solve_cubic_eq(VanDerWaalsEos._zfactor_cubic_eq(A, B))

    @staticmethod
    def ln_fugacity_coeff(z, state):
        """
        Computes the natural log of fugacity coefficient

        Parameters
        ----------
        z : array_like
            Z-factors
        state : IsobaricIsothermalState

        Returns
        -------
        array_like
            Natural log of fugacity coefficients
        """
        A = state.reduced_attraction_param
        B = state.reduced_repulsion_param
        return VanDerWaalsEos._ln_fugacity_coeff_impl(z, A, B)

    @staticmethod
    def fugacity_coeff(z, state):
        """
        Computes fugacity coefficient

        Parameters
        ----------
        z : array_like
            Z-factors
        state : IsobaricIsothermalState

        Returns
        -------
        array_like
            Fugacity coefficients
        """
        return np.exp(VanDerWaalsEos.ln_fugacity_coeff(z, state))

File: matlab/+eos/VanDerWaalsEos.m
classdef VanDerWaalsEos < eos.CubicEosBase
    % VanDerWaalsEos Van der Waals equation of state
    %
    %  This class provides methods to calculate thermodynamic properties
    %  based on Van der Waals equation of state.
    
    methods (Static)
        function coeffs = zFactorCubicEq(A,B)
            % Computes coefficients of Z-factor cubic equation
            %
            % Parameters
            % ----------
            % A : Reduced attraction parameter
            % B : Reduced repulsion parameter
            %
            % Returns
            % -------
            % coeffs : Coefficients of the cubic equation of Z-factor
            coeffs = [1, -B - 1, A, -A*B];
        end
        function lnPhi = lnFugacityCoeff(z,A,B)
            % Computes natural log of fugacity coefficients
            %
            % Parameters
            % ----------
            % z : Z-factor
            % A : Reduced attraction parameter
            % B : Reduced repulsion parameter
            %
            % Returns
            % -------
            % lnPhi : Natural log of fugacity coefficients
            lnPhi = z - 1 - log(z - B) - A./z;
        end
        function phi = fugacityCoeff(z,A,B)
            % Computes fugacity coefficients
            %
            % Parameters
            % ----------
            % z : Z-factor
            % A : Reduced attraction parameter
            % B : Reduced repulsion parameter
            %
            % Returns
            % -------
            % phi : Fugacity coefficients
            phi = exp(eos.VanDerWaalsEos.lnFugacityCoeff(z,A,B));
        end
    end
    methods
        function obj = VanDerWaalsEos(Pc,Tc,Mw)
            % Constructs VDW EOS
            %
            % Parameters
            % ----------
            % Pc : Critical pressure [Pa]
            % Tc : Critical temperature [K]
            % Mw : Molecular weight [g/mol]
            obj@eos.CubicEosBase(0.421875,0.125,Pc,Tc,Mw);
        end
        function obj = setParams(obj,Pc,Tc,Mw)
            % Set parameters
            %
            % Parameters
            % ----------
            % Pc : Critical pressure [Pa]
            % Tc : Critical temperature [K]
            % Mw : Molecular weight [g/mol]
            obj = setParams@eos.CubicEosBase(obj,Pc,Tc,Mw);
        end
        function P = pressure(obj,T,V)
            % Computes pressure
            %
            % Parameters
            % ----------
            % T : Temperature [K]
            % V : Volume [m3]
            %
            % Returns
            % -------
            % P : Pressure [Pa]
            R = eos.ThermodynamicConstants.Gas;
            a = obj.AttractionParam;
            b = obj.RepulsionParam;
            P = R*T./(V - b) - a./V^2;
        end
        function [z,A,B] = zFactors(obj,P,T)
            % Computes Z-factors
            %
            % Parameters
            % ----------
            % P : Pressure [Pa]
            % T : Temperature [K]
            %
            % Returns
            % -------
            % z : Z-factors
            % A : Reduced attraction parameter
            % B : Reduced repulsion parameter
            Pr = obj.reducedPressure(P);
            Tr = obj.reducedTemperature(T);
            A = obj.reducedAttractionParam(Pr,Tr,1);
            B = obj.reducedRepulsionParam(Pr,Tr);
            x = roots(obj.zFactorCubicEq(A,B));
            z = x(imag(x) == 0);
        end
    end
end